src/tools/rbd_mirror/image_replayer/snapshot/Replayer.cc

```diff
@@ -418,32 +419,46 @@ void Replayer<I>::scan_local_mirror_snapshots(
              << "mirror_ns=" << *mirror_ns << dendl;
     m_local_mirror_snap_ns = *mirror_ns;
 
+    bool completed_non_primary_snapshots_exist = false;
     auto local_snap_id = snap_info_it->first;
     if (mirror_ns->is_non_primary()) {
       if (mirror_ns->complete) {
         // if remote has new snapshots, we would sync from here
         m_local_snap_id_start = local_snap_id;
         m_local_snap_id_end = CEPH_NOSNAP;
+        completed_non_primary_snapshots_exist = true;
 
         if (mirror_ns->mirror_peer_uuids.empty()) {
           // no other peer will attempt to sync to this snapshot so store as
           // a candidate for removal
           prune_snap_ids.insert(local_snap_id);
         }
-      } else {
-        if (mirror_ns->last_copied_object_number == 0) {
-          // snapshot might be missing image state, object-map, etc, so just
-          // delete and re-create it if we haven't started copying data
-          // objects. Also only prune this snapshot since we will need the
-          // previous mirror snapshot for syncing.
-          prune_snap_ids.clear();
-          prune_snap_ids.insert(local_snap_id);
-          break;
+      } else if (mirror_ns->last_copied_object_number == 0 &&
+                 m_local_snap_id_start > 0) {
+        // shouldn't be possible, but ensure that pruning this snapshot
+        // wouldn't leave this image w/o any non-primary snapshots
+        if (!completed_non_primary_snapshots_exist) {
+          derr << "incomplete local non-primary snapshot" << dendl;
+          handle_replay_complete(locker, -EINVAL,
+                                 "incomplete local non-primary snapshot");
+          return;
```

Comment:
@dillaman observing mirror snapshot stress test failures due to a replayer failed on start with "incomplete local non-primary snapshot" [1].

I looked at one case [2]. It is observed after an image is resumed on one instance and acquired on another. And when it is being resumed, it is in process of replaying a snapshot: creates the destination snapshot but before starting copying detects that shutdown is requested leaving incomplete snapshot.
```
cluster1-client.mirror.1.44130.log.gz:

2021-02-15T10:54:22.388+0000 7fc72fa5f700 10 rbd::mirror::image_replayer::snapshot::Replayer: 0x559a87d6f000 shut_down: 
2021-02-15T10:54:22.388+0000 7fc72fa5f700 10 rbd::mirror::InstanceWatcher: 0x559a81d24a80 cancel_sync_request: sync_id=12c925a0ebfc
2021-02-15T10:54:22.388+0000 7fc72fa5f700 10 rbd::mirror::image_replayer::snapshot::Replayer: 0x559a87d6f000 shut_down: shut down pending on completion of snapshot replay
....
2021-02-15T10:54:23.490+0000 7fc729252700 10 rbd::mirror::image_replayer::snapshot::Replayer: 0x559a87d6f000 handle_create_non_primary_snapshot: r=0
2021-02-15T10:54:23.490+0000 7fc729252700 15 rbd::mirror::image_replayer::snapshot::Replayer: 0x559a87d6f000 handle_create_non_primary_snapshot: local_snap_id_end=239
2021-02-15T10:54:23.490+0000 7fc729252700 10 rbd::mirror::image_replayer::snapshot::Replayer: 0x559a87d6f000 request_sync: 
2021-02-15T10:54:23.490+0000 7fc729252700 10 rbd::mirror::image_replayer::snapshot::Replayer: 0x559a87d6f000 is_replay_interrupted: resuming pending shut down
2021-02-15T10:54:23.490+0000 7fc729252700 10 rbd::mirror::image_replayer::snapshot::Replayer: 0x559a87d6f000 unregister_remote_update_watcher: 
2021-02-15T10:54:23.490+0000 7fc729252700 10 rbd::mirror::image_replayer::snapshot::Replayer: 0x559a87d6f000 handle_unregister_remote_update_watcher: r=0
2021-02-15T10:54:23.490+0000 7fc729252700 10 rbd::mirror::image_replayer::snapshot::Replayer: 0x559a87d6f000 unregister_local_update_watcher: 
2021-02-15T10:54:23.490+0000 7fc729252700 10 rbd::mirror::image_replayer::snapshot::Replayer: 0x559a87d6f000 handle_unregister_local_update_watcher: r=0
2021-02-15T10:54:23.490+0000 7fc729252700 10 rbd::mirror::image_replayer::snapshot::Replayer: 0x559a87d6f000 wait_for_in_flight_ops: 
2021-02-15T10:54:23.490+0000 7fc72f25e700 10 rbd::mirror::image_replayer::snapshot::Replayer: 0x559a87d6f000 handle_wait_for_in_flight_ops: r=0
2021-02-15T10:54:23.490+0000 7fc72f25e700 10 rbd::mirror::image_replayer::snapshot::Replayer: 0x559a87d6f000 ~Replayer: 
```

And then when another instance starts the replayer, it fails due to the snapshot incomplete:
```
cluster1-client.mirror.1.55775.log.gz:

2021-02-15T10:54:51.553+0000 7f389548f700 10 rbd::mirror::InstanceReplayer: 0x55baf4aec280 start_image_replayer: global_image_id=ab31b979-007c-49bc-8088-544ae5494ad1
2021-02-15T10:54:51.553+0000 7f389548f700 10 rbd::mirror::ImageReplayer: 0x55baf66c2280 [3/ab31b979-007c-49bc-8088-544ae5494ad1] start: on_finish=0x55baf667b1e0
...
2021-02-15T10:54:51.659+0000 7f388ec82700 15 rbd::mirror::image_replayer::snapshot::Replayer: 0x55baf6d2b400 scan_local_mirror_snapshots: local mirror snapshot: id=107, mirror_ns=[mirror state=non-primary, complete=1, mirror_peer_uuids=, primary_mirror_uuid=479e16be-4a87-4fd4-b053-c75a5ca44436, primary_snap_id=6c, last_copied_object_number=0, snap_seqs={108=18446744073709551614}]
2021-02-15T10:54:51.659+0000 7f388ec82700 15 rbd::mirror::image_replayer::snapshot::Replayer: 0x55baf6d2b400 scan_local_mirror_snapshots: local mirror snapshot: id=239, mirror_ns=[mirror state=non-primary, complete=0, mirror_peer_uuids=, primary_mirror_uuid=479e16be-4a87-4fd4-b053-c75a5ca44436, primary_snap_id=ef, last_copied_object_number=0, snap_seqs={238=238,239=18446744073709551614}]
2021-02-15T10:54:51.659+0000 7f388ec82700 -1 rbd::mirror::image_replayer::snapshot::Replayer: 0x55baf6d2b400 scan_local_mirror_snapshots: incomplete local non-primary snapshot
2021-02-15T10:54:51.659+0000 7f388ec82700 10 rbd::mirror::image_replayer::snapshot::Replayer: 0x55baf6d2b400 notify_status_updated: 
2021-02-15T10:54:51.659+0000 7f389548f700 10 rbd::mirror::ImageReplayer: 0x55baf66c2280 [3/ab31b979-007c-49bc-8088-544ae5494ad1] handle_replayer_notification: 
2021-02-15T10:54:51.659+0000 7f389548f700 10 rbd::mirror::ImageReplayer: 0x55baf66c2280 [3/ab31b979-007c-49bc-8088-544ae5494ad1] handle_replayer_notification: replay interrupted: r=-22, error=incomplete local non-primary snapshot
```

[1] https://pulpito.ceph.com/trociny-2021-02-15_09:32:56-rbd-wip-mgolub-testing-distro-basic-smithi/
[2] http://qa-proxy.ceph.com/teuthology/trociny-2021-02-15_09:32:56-rbd-wip-mgolub-testing-distro-basic-smithi/5883923/teuthology.log

Reply:
Oops -- `completed_non_primary_snapshots_exist` was incorrectly being re-initialized to false within the loop.